lib/puppet_x/puppetlabs/registry.rb

```diff
@@ -124,45 +151,79 @@ def filter_path
   end
 
   class RegistryKeyPath < RegistryPathBase
-    def subkey
-      filter_path[:trailing_path]
-    end
   end
 
   class RegistryValuePath < RegistryPathBase
-    def canonical
-      # This method gets called in the type and the provider.  We need to
-      # preserve the trailing backslash for the provider, otherwise it won't
-      # think this is a default value.
-      if default?
-        filter_path[:canonical] + "\\"
+    attr_reader :valuename
+
+    # Combines a registry key path and valuename into a resource title for
+    # registry_value resource.
+    #
+    # To maintain backwards compatibility, only use the double backslash
+    # delimiter if the valuename actually contains a backslash
+    def self.combine_path_and_value(keypath, valuename)
+      if valuename.include?('\\')
+        keypath + '\\\\' + valuename
       else
-        filter_path[:canonical]
+        keypath + '\\' + valuename
       end
     end
 
-    def subkey
-      if default?
-        filter_path[:trailing_path]
+
+    # Extract the valuename from the path and then munge the actual path
+    def initialize(path)
+      # Try finding the valuename via the double backslash method first
+      # and then revert to the old single backslash way
+      dbl_slash_idx = path.index('\\\\') || 0
```

Comment:
I think all of the array index logic can be simplified by setting the limit param to 2 on `split`, and a destructuring assignment like this:

```
[1] pry(main)> path, valuename = "foo\\\\bar".split("\\\\", 2)
=> ["foo", "bar"]
[2] pry(main)> path
=> "foo"
[3] pry(main)> valuename
=> "bar"

[4] pry(main)> path, valuename = "foo\\\\bar\\\\".split("\\\\", 2)
=> ["foo", "bar\\\\"]
[5] pry(main)> path
=> "foo"
[6] pry(main)> valuename
=> "bar\\\\"

[7] pry(main)> path, valuename = "foo\\\\\\bar\\".split("\\\\", 2)
=> ["foo", "\\bar\\"]
[8] pry(main)> path
=> "foo"
[9] pry(main)> valuename
=> "\\bar\\"

[10] pry(main)> path, valuename = "foo\\\\\\bar\\baz\\\\".split("\\\\", 2)
=> ["foo", "\\bar\\baz\\\\"]
[11] pry(main)> path
=> "foo"
[12] pry(main)> valuename
=> "\\bar\\baz\\\\"

[13] pry(main)> path, valuename = "foo\\bar\\".split("\\\\", 2)
=> ["foo\\bar\\"]
[14] pry(main)> path
=> "foo\\bar\\"
[15] pry(main)> valuename
=> nil

[16] pry(main)> path, valuename = "foobar".split("\\\\", 2)
=> ["foobar"]
[17] pry(main)> path
=> "foobar"
[18] pry(main)> valuename
=> nil
```

With no double backslash, `valuename` is `nil`, in which case you can fall back to doing the same thing with the single backslash..

Also, `end_with?('\\')` is a bit easier to read than a case statement against `path[-1, 1]`